prototypes/coin-selection/CoinSelection.hs

```diff
@@ -0,0 +1,278 @@
+data Selector
+    = Sweep
+    -- ^ Use all possible unspent transaction inputs, regardless of the
+    -- requested payment amount. This is called a wallet "sweep" -- where all
+    -- coins from the wallet are shifted together.
+    | RandomImprove
+    -- ^ Use the /Random-Improve-MA/ algorithm for selecting inputs. This will
+    -- select approximately double the amount necessary for the payment.
+    deriving (Show, Read, Eq, Enum, Generic)
+
+-- | How to assign excess value to change outputs.
+data Change
+    = Balanced
+    -- ^ Value is distributed evenly between change outputs.
+    | Proportional
+    -- ^ Value is distributed between change outputs in proportion to payment
+    -- outputs. Use this with v'RandomImprove' for /emergent/ management of the
+    -- wallet's UTxO.
+    | Unbalanced
+    -- ^ Value is added to the first change output until it reaches the maximum
+    -- size. Then the next change output will receive value.
+    deriving (Show, Read, Eq, Enum, Generic)
+
+-- | Placeholder for the type of possible failure cases in coin selection.
+newtype SelectionError = SelectionError String -- ^ Something went wrong.
+    deriving (Show, Read, Eq, Generic)
+
+-- $naming
+-- We call the coin selection function 'xyzzy' in lieu of a better name.
+--
+-- Possible names could be @balanceTransaction@, @selectCoins@, or
+-- something like that.
+
+-- | The top-level coin selection function.
+--
+-- Given an initial partially funded transaction body, it will produce a
+-- complete, funded, and balanced transaction body from the provided 'UTxO'.
+--
+-- If successful, it returns one or more transactions, otherwise the reason for
+-- failure is provided.
+--
+-- __How to use__
+--
+-- The main interface of 'xyzzy' is its parameter @tx@ the partial
+-- transaction. The caller specifies the outputs of @tx@, and 'xyzzy' calculates
+-- a balanced 'TxBody' from that.
+--
+-- [Payment]: Add a 'TxOut' to the initial @tx@ with the recipient
+--   'Address' and 'Value'.
+-- [Change Output]: Add a 'TxOut' to the initial @tx@ such that the 'Address'
+--   satisfies 'isOurAddress'. Its 'Value' would usually be zero.
+--   If the maximum output size is reached, another 'TxOut' with the same
+--   address will be added to the transaction.
+-- [No new change addresses]: If there is no transaction output that satisfies
+--   v'IsOurAddress', then change will be directed back to the addresses
+--   belonging to the wallet which funded the transaction.
+-- [Join stake pool]: This transaction will involve a delegation certificate,
+--   and possibly also a stake key registration certificate. These cost money.
+--   However 'xyzzy', is not necessarily Cardano-specific or even POS-specific.
+--   Therefore, the 'Address' type for @tx@ should include a constructor for
+--   certificates. The implementation of 'appendTxOutput' should put these
+--   "certificate" address outputs in the correct place within the concrete
+--   'TxBody' type.
```

Comment:
The abstract coin-selection shouldn't need to know about certificates. They can be treated as something which affects the size (and therefore fee) as well as balance (w.r.t deposits) of a transaction via some `evaluateTransaction{Balance,MinFee}` functions passed to `xyzzy`.